src/metabase/middleware.clj

```diff
@@ -247,15 +242,28 @@
     "X-Content-Type-Options"            "nosniff"}))
 
 (defn add-security-headers
-  "Add HTTP headers to tell browsers not to cache API responses."
+  "Add HTTP security and cache-busting headers."
+  [handler]
+  (fn [request]
+    (let [response (handler request)]
+      ;; add security headers to all responses, but allow iframes on public & embed responses
+      (update response :headers merge (security-headers :allow-iframes? ((some-fn public? embed?) request))))))
+
+(defn add-content-type
+  "Add an appropriate Content-Type header to response if it doesn't already have one. Most responses should already
+  have one, so this is a fallback for ones that for one reason or another do not."
   [handler]
   (fn [request]
     (let [response (handler request)]
-      (update response :headers merge (cond
-                                        (api-call? request) (api-security-headers)
-                                        (public? request)   (html-page-security-headers, :allow-iframes? true)
-                                        (embed? request)    (html-page-security-headers, :allow-iframes? true)
-                                        (index? request)    (html-page-security-headers))))))
+      (update-in
+       response
+       [:headers "Content-Type"]
+       (fn [content-type]
+         (or content-type
+             (when (api-call? request)
+               (if (string? (:body response))
```

Comment:
Do we not have any places in the code where we return a JSON string (and thus would mislabel it as `text/plain`)?

Reply:
I don’t believe we do
But if we do it should almost certainly already have a header set